protoarray/src/main/java/tech.pegasys.artemis.protoarray/ProtoArray.java

```diff
@@ -0,0 +1,380 @@
+          .ifPresent(
+              bestDescendantIndex -> {
+                int newBestDescendantIndex = bestDescendantIndex - finalizedIndex;
+                if (newBestDescendantIndex < 0) {
+                  throw new RuntimeException(
+                      "ProtoArray: New best descendant index is less than 0");
+                }
+                node.setBestDescendantIndex(Optional.of(newBestDescendantIndex));
+              });
+    }
+  }
+
+  /**
+   * Observe the parent at `parentIndex` with respect to the child at `childIndex` and
+   * potentially modify the `parent.bestChild` and `parent.bestDescendant` values.
+   *
+   * ## Detail
+   *
+   * There are four outcomes:
+   *
+   * - The child is already the best child but it's now invalid due to a FFG change and should be
+   * removed.
+   * - The child is already the best child and the parent is updated with the new best descendant.
+   * - The child is not the best child but becomes the best child.
+   * - The child is not the best child and does not become the best child.
+   *
+   * @param parentIndex
+   * @param childIndex
+   */
+  private void maybeUpdateBestChildAndDescendant(int parentIndex, int childIndex) {
+    ProtoNode child = nodes.get(childIndex);
+    ProtoNode parent = nodes.get(parentIndex);
+
+    boolean childLeadsToViableHead = nodeLeadsToViableHead(child);
+
+    parent
+        .getBestChildIndex()
+        .ifPresentOrElse(
+            bestChildIndex -> {
+              if (bestChildIndex.equals(childIndex) && !childLeadsToViableHead) {
+                // If the child is already the best-child of the parent but it's not viable for
+                // the head, remove it.
+                changeToNone(parent);
+              } else if (bestChildIndex.equals(childIndex)) {
+                // If the child is the best-child already, set it again to ensure that the
+                // best-descendant of the parent is updated.
+                changeToChild(parent, childIndex);
+              } else {
+                ProtoNode bestChild = nodes.get(bestChildIndex);
+
+                boolean bestChildLeadsToViableHead = nodeLeadsToViableHead(bestChild);
+
+                if (childLeadsToViableHead && !bestChildLeadsToViableHead) {
+                  // The child leads to a viable head, but the current best-child doesn't.
+                  changeToChild(parent, childIndex);
+                } else if (!childLeadsToViableHead && bestChildLeadsToViableHead) {
+                  // The best child leads to a viable head, but the child doesn't.
+                  // No change.
+                } else if (child.getWeight().equals(bestChild.getWeight())) {
+                  // Tie-breaker of equal weights by root.
+                  if (child.getRoot().compareTo(bestChild.getRoot()) >= 0) {
+                    changeToChild(parent, childIndex);
+                  } else {
+                    // No change.
+                  }
```

Comment:
Would be worth adding `
  @SuppressWarnings("StatementWithEmptyBody")` to the method so IntelliJ doesn't warn about these empty blocks.

Reply:
Good call. Done.